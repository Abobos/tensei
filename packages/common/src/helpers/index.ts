import { Request } from 'express'
import { ResourceContract } from '@tensei/common'

export class ResourceHelpers {
    public resource: ResourceContract | null = null

    constructor(public resources: ResourceContract[]) {}

    public findResource = (resourceSlug: string | ResourceContract) => {
        if (!resourceSlug) {
            throw {
                message: `Resource ${resourceSlug} not found.`,
                status: 404
            }
        }

        if (typeof resourceSlug !== 'string') {
            return resourceSlug
        }

        const resource = this.resources.find(
            resource =>
                resource.data.slug === resourceSlug ||
                resource.data.name === resourceSlug
        )

        if (!resource) {
            throw {
                message: `Resource ${resourceSlug} not found.`,
                status: 404
            }
        }

        return resource
    }

    getFieldFromResource = (
        resource: ResourceContract,
        databaseField: string
    ) => {
        return resource.data.fields.find(
            field =>
                field.name === databaseField ||
                field.databaseField === databaseField
        )
    }

    setResource = (resourceOrSlug: ResourceContract | string) => {
        this.resource = this.findResource(resourceOrSlug)

        return this
    }

    protected getCurrentResource = () => {
        if (!this.resource) {
            throw {
                message: `Missing resource for this operation.`,
                status: 400
            }
        }

        return this.resource
    }
}

/**
 * Apply typescript mixins to a class constructor.
 *
 * @param derivedCtor any
 * @param constructors any
 */
export const mix = (derivedCtor: any, constructors: any[]) => {
    constructors.forEach(baseCtor => {
        Object.getOwnPropertyNames(baseCtor.prototype).forEach(name => {
            const propertyDescription = Object.getOwnPropertyDescriptor(
                baseCtor.prototype,
                name
            )

            if (!propertyDescription) {
                return
            }

            Object.defineProperty(
                derivedCtor.prototype,
                name,
                propertyDescription
            )
        })
    })
}
