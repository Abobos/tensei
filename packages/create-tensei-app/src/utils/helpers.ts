import Fs from 'fs'
import Ora from 'ora'
import Path from 'path'
import Execa from 'execa'
import Edge from 'edge.js'
import latestVersion from 'latest-version'
import signale from './signale'

export const getAvailableTemplates = () => ['quickstart']

export const getAvailableDatabases = () => ['sqlite', 'mysql', 'postgresql']

export const getPackageVersions = (database: string, api: string) =>
    Promise.all([
        latestVersion(`@mikro-orm/${database}`),
        latestVersion(`@tensei/core`),
        latestVersion(`@tensei/auth`),
        latestVersion(`@tensei/${api}`),
        latestVersion(`nodemon`)
    ])

export const parseTemplateFiles = (
    template: string,
    templateVariables: any
) => {
    const basePath = Path.resolve(__dirname, '..', '..', 'templates', template)

    return Fs.readdirSync(basePath).map(file => [
        file,
        Edge.renderString(
            Fs.readFileSync(Path.resolve(basePath, file)).toString(),
            templateVariables
        )
    ])
}

export const projectDirectoryExists = (project: string) =>
    Fs.existsSync(Path.resolve(process.cwd(), project))

export const createProjectFolder = (project: string) =>
    Fs.mkdirSync(Path.resolve(process.cwd(), project))

export const createProjectFile = (
    project: string,
    file: string,
    content: string
) =>
    Fs.writeFileSync(
        Path.resolve(
            process.cwd(),
            project,
            file.endsWith('.edge') ? file.substring(0, file.length - 5) : file
        ),
        content
    )

export const hasYarnInstalled = () => {
    try {
        return Execa.commandSync('yarn --version').exitCode === 0
    } catch (e) {
        return false
    }
}

export const installProjectDependencies = (
    packageManager: string,
    project: string,
    runDev = true
) => {
    try {
        const spinner = Ora(
            `Installing dependencies with ${packageManager}`
        ).start()

        Execa.sync(packageManager, ['install'], {
            cwd: Path.resolve(process.cwd(), project),
            stdin: 'ignore'
        })

        spinner.stop()

        signale.success(
            `ðŸš€ Done installing project.\n\nðŸŽ‰ Congratulations on the start of something amazing!\n\nðŸ’» Run the following command to get started:\n\ncd ${project} && ${packageManager} run dev\n\n`
        )

        process.exit(0)
    } catch (e) {
        signale.error(e)

        process.exit(1)
    }
}
