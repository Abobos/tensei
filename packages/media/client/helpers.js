const units = ['bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']

const imageExtensions = [
    'ase',
    'art',
    'bmp',
    'blp',
    'cd5',
    'cit',
    'cpt',
    'cr2',
    'cut',
    'dds',
    'dib',
    'djvu',
    'egt',
    'exif',
    'gif',
    'gpl',
    'grf',
    'icns',
    'ico',
    'iff',
    'jng',
    'jpeg',
    'jpg',
    'jfif',
    'jp2',
    'jps',
    'lbm',
    'max',
    'miff',
    'mng',
    'msp',
    'nitf',
    'ota',
    'pbm',
    'pc1',
    'pc2',
    'pc3',
    'pcf',
    'pcx',
    'pdn',
    'pgm',
    'PI1',
    'PI2',
    'PI3',
    'pict',
    'pct',
    'pnm',
    'pns',
    'ppm',
    'psb',
    'psd',
    'pdd',
    'psp',
    'px',
    'pxm',
    'pxr',
    'qfx',
    'raw',
    'rle',
    'sct',
    'sgi',
    'rgb',
    'int',
    'bw',
    'tga',
    'tiff',
    'tif',
    'vtf',
    'xbm',
    'xcf',
    'xpm',
    '3dv',
    'amf',
    'ai',
    'awg',
    'cgm',
    'cdr',
    'cmx',
    'dxf',
    'e2d',
    'egt',
    'eps',
    'fs',
    'gbr',
    'odg',
    'svg',
    'stl',
    'vrml',
    'x3d',
    'sxd',
    'v2d',
    'vnd',
    'wmf',
    'emf',
    'art',
    'xar',
    'png',
    'webp',
    'jxr',
    'hdp',
    'wdp',
    'cur',
    'ecw',
    'iff',
    'lbm',
    'liff',
    'nrrd',
    'pam',
    'pcx',
    'pgf',
    'sgi',
    'rgb',
    'rgba',
    'bw',
    'int',
    'inta',
    'sid',
    'ras',
    'sun',
    'tga'
]

export const niceBytes = x => {
    let l = 0,
        n = parseInt(x, 10) || 0

    while (n >= 1024 && ++l) {
        n = n / 1024
    }

    return n.toFixed(n < 10 && l > 0 ? 1 : 0) + ' ' + units[l]
}

export const getSecondSmallestSizeHash = asset => {
    const transforms = asset.transformations.sort(
        (tA, tB) => tA.width - tB.width
    )

    if (transforms.length >= 2) {
        return transforms[1].hash
    }

    if (transforms.length === 1) {
        return transforms[0].hash
    }

    return asset.hash
}

export const getFileType = file => {
    if (isImage(file)) {
        return 'IMAGE'
    }

    return file.extension
}

export const isImage = asset => imageExtensions.includes(asset.extension)

export const getFullAssetPath = asset => {
    return `${window.Tensei.state.config.serverUrl}${
        asset.path
    }${getSecondSmallestSizeHash(asset)}.${asset.extension}`
}
