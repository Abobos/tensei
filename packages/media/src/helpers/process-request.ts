import Busboy from 'busboy'
import Crypto from 'crypto'
import createError from 'http-errors'
import { WriteStream } from 'fs-capacitor'
import { Request, Response } from 'express'
import { DataPayload, ApiContext } from '@tensei/common'
import { MediaLibraryPluginConfig, UploadFile } from '../types'

import { mediaResource } from '../resources'

const ignoreStream = (stream: NodeJS.ReadableStream) => {
    stream.on('error', () => {})

    stream.resume()
}

export const process = (
    request: Request,
    response: Response,
    config: MediaLibraryPluginConfig
) => {
    return new Promise((resolve, reject) => {
        const parser = new Busboy({
            headers: request.headers,
            limits: {
                fieldSize: config.maxFieldSize,
                fields: 2, // Only operations and map.
                fileSize: config.maxFileSize,
                files: config.maxFiles
            }
        })

        let exitError: Error | undefined
        let uploads: any[] = []
        let body: DataPayload = {}
        let released: boolean | undefined
        let currentStream: NodeJS.ReadableStream | undefined

        const exit = (error: Error) => {
            if (exitError) return
            exitError = error

            reject(exitError)

            // @ts-ignore
            if (parser?.destroy) {
                // @ts-ignore
                parser.destroy()
            }

            if (currentStream) {
                // @ts-ignore
                if (currentStream.destroy) {
                    // @ts-ignore
                    currentStream.destroy(exitError)
                }
            }

            request.unpipe(parser)

            // With a sufficiently large request body, subsequent events in the same
            // event frame cause the stream to pause after the parser is destroyed. To
            // ensure that the request resumes, the call to .resume() is scheduled for
            // later in the event loop.
            setImmediate(() => {
                request.resume()
            })
        }

        const release = () => {
            if (released) return
            released = true

            uploads.forEach(upload => upload.file?.capacitor?.release())
        }

        const abort = () => {
            exit(
                createError(
                    499,
                    'Request disconnected during file upload stream parsing.'
                )
            )
        }

        parser.on('field', (fieldname, value, fieldnameTruncated) => {
            if (fieldnameTruncated) {
                return exit(
                    createError(
                        413,
                        `The ‘${fieldname}’ multipart field value exceeds the ${config.maxFieldSize} byte size limit.`
                    )
                )
            }

            body[fieldname] = value
        })

        parser.on('file', (fieldName, stream, filename, encoding, mimetype) => {
            if (exitError) {
                ignoreStream(stream)

                return
            }

            let fileError: Error | undefined

            currentStream = stream
            stream.on('end', () => {
                currentStream = undefined
            })

            const capacitor = new WriteStream()

            capacitor.on('error', () => {
                stream.unpipe()
                stream.resume()
            })

            stream.on('limit', () => {
                fileError = createError(
                    413,
                    `File truncated as it exceeds the ${config.maxFileSize} byte size limit.`
                )
                stream.unpipe()
                capacitor.destroy(fileError)
            })

            stream.on('error', error => {
                fileError = error
                stream.unpipe()
                capacitor.destroy(exitError)
            })

            const file = {
                filename,
                mimetype,
                encoding,
                fieldName,
                createReadStream(name?: string) {
                    const error = fileError || (released ? exitError : null)

                    if (error) throw error

                    return capacitor.createReadStream(name)
                }
            }

            Object.defineProperty(file, 'capacitor', { value: capacitor })

            stream.pipe(capacitor)
            uploads.push(file)
        })

        parser.once('filesLimit', () =>
            exit(
                createError(
                    413,
                    `${config.maxFiles} max file uploads exceeded.`
                )
            )
        )

        parser.once('finish', () => {
            request.unpipe(parser)
            request.resume()

            resolve({
                files: uploads,
                ...body
            })
        })

        parser.once('error', exit)

        response.once('finish', release)
        response.once('close', release)

        request.once('close', abort)
        request.once('end', () => {
            request.removeListener('close', abort)
        })

        request.pipe(parser)
    })
}

export const handle = async (
    ctx: ApiContext,
    config: MediaLibraryPluginConfig
) => {
    let files = ((await Promise.all(ctx.body?.files)) as UploadFile[]).map(
        file => {
            const split = file.filename.split('.')
            const extension = split[split.length - 1]

            const hash = Crypto.randomBytes(36).toString('hex')

            let file_path: string = ctx.body.path || '/'

            file_path = file_path.startsWith('/') ? file_path : `/${file_path}`
            file_path = file_path.endsWith('/') ? file_path : `${file_path}/`

            return {
                ...file,
                hash,
                extension,
                path: file_path,
                storage_filename: `${file_path}${hash}.${extension}`
            }
        }
    )

    await Promise.all(
        files.map(file =>
            ctx.storage
                .disk(config.disk)
                .put(`${file.storage_filename}`, file.createReadStream())
        )
    )

    const storedFiles = await Promise.all(
        files.map(file => ctx.storage.disk().getStat(file.storage_filename))
    )

    files = files.map((file, index) => ({
        ...file,
        size: storedFiles[index].size
    }))

    const entities = files.map(file =>
        ctx.manager.create(mediaResource().data.pascalCaseName, {
            size: file.size,
            hash: file.hash,
            path: file.path,
            disk: config.disk,
            mime_type: file.mimetype,
            extension: file.extension,
            original_filename: file.filename
        })
    )

    await ctx.manager.persistAndFlush(entities)

    return entities
}
