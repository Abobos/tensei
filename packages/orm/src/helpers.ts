import { ReferenceType } from '@mikro-orm/core'
import { FieldContract, ResourceContract } from '@tensei/common'

export const resolveFieldTypescriptType = (
    field: FieldContract,
    resources: ResourceContract[]
  ) => {
    if (
      ['integer', 'bigInteger', 'int', 'number', 'float', 'double'].includes(
        field.property.type!
      )
    ) {
      return 'number'
    }
  
    if (field.property.type === 'boolean') {
      return 'boolean'
    }
  
    if (field.property.type === 'date') {
      return 'Date'
    }
  
    if (field.property.type === 'json') {
      if (field.validationRules.includes('array')) {
        if (field.arrayValidationRules.includes('string')) {
          return 'string[]'
        }
  
        if (field.arrayValidationRules.includes('number')) {
          return 'number[]'
        }
  
        if (field.arrayValidationRules.includes('decimal')) {
          return 'number[]'
        }
  
        if (field.arrayValidationRules.includes('date')) {
          return 'Date[]'
        }
      }
    }
  
    if (
      field.relatedProperty.reference === ReferenceType.MANY_TO_ONE ||
      field.relatedProperty.reference === ReferenceType.ONE_TO_ONE
    ) {
      const relatedResource = resources.find(
        resource => resource.data.name === field.name
      )
  
      if (relatedResource) {
        return `${relatedResource.data.pascalCaseName}Model`
      }
  
      return 'ID'
    }
  
    if (
      field.relatedProperty.reference === ReferenceType.ONE_TO_MANY ||
      field.relatedProperty.reference === ReferenceType.MANY_TO_MANY
    ) {
      const relatedResource = resources.find(
        resource => resource.data.name === field.name
      )
  
      if (relatedResource) {
        return `${relatedResource.data.pascalCaseName}Model[]`
      }
  
      return 'ID[]'
    }
  
    return 'string'
  }
