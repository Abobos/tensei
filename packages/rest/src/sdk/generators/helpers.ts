import Prettier from 'prettier'
import { ReferenceType } from '@mikro-orm/core'
import { FieldContract, ResourceContract } from '@tensei/common'

export function formatContent(content: string) {
    return Prettier.format(content, {
        semi: false,
        parser: 'typescript',
        singleQuote: true
    })
}

export const resolveFieldTypescriptType = (
    field: FieldContract,
    resources: ResourceContract[],
    namespaced = false
) => {
    if (
        ['integer', 'bigInteger', 'int', 'number', 'float', 'double'].includes(
            field.property.type!
        )
    ) {
        return 'number'
    }

    if (field.property.type === 'boolean') {
        return 'boolean'
    }

    if (field.property.type === 'date') {
        return 'DateString'
    }

    if (field.property.type === 'json') {
        if (field.validationRules.includes('array')) {
            if (field.arrayValidationRules.includes('string')) {
                return 'string[]'
            }

            if (field.arrayValidationRules.includes('number')) {
                return 'number[]'
            }

            if (field.arrayValidationRules.includes('decimal')) {
                return 'number[]'
            }

            if (field.arrayValidationRules.includes('date')) {
                return 'DateString[]'
            }
        }
    }

    if (
        field.relatedProperty.reference === ReferenceType.MANY_TO_ONE ||
        field.relatedProperty.reference === ReferenceType.ONE_TO_ONE
    ) {
        const relatedResource = resources.find(
            resource => resource.data.name === field.name
        )

        if (relatedResource) {
            return `${relatedResource.data.pascalCaseName}['id']`
        }

        return 'ID'
    }

    if (
        field.relatedProperty.reference === ReferenceType.ONE_TO_MANY ||
        field.relatedProperty.reference === ReferenceType.MANY_TO_MANY
    ) {
        const relatedResource = resources.find(
            resource => resource.data.name === field.name
        )

        if (relatedResource) {
            return `${relatedResource.data.pascalCaseName}['id'][]`
        }

        return 'ID[]'
    }

    return 'string'
}
