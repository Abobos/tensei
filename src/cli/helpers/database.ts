import Path from 'path'
import Mongo from 'mongodb'
import Dotenv from 'dotenv'

export default async () => {
    Dotenv.config({
        path: Path.join(process.env.PWD!, '.env'),
    })

    const client = new Mongo.MongoClient('mongodb://localhost/flamingo', {
        useNewUrlParser: true,
        useUnifiedTopology: true,
    })

    await client.connect()

    return client
}
