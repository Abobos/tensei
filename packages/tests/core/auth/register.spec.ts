import Path from 'path'
import Knex from 'knex'
import Supertest from 'supertest'
import isAfter from 'date-fns/isAfter'

import { setup, createAdminUser } from '../../helpers'

['mysql', 'sqlite3'].forEach((databaseClient: any) => {
    test(`${databaseClient} -  validates registration data and returns error messages with a 422`, async () => {
        const { app } = await setup({
            databaseClient
        })
    
        const client = Supertest(app)
    
        const response = await client.post('/api/register').send({})
    
        expect(response.status).toBe(422)
        expect(response.body).toMatchSnapshot()
    })
    
    test(`${databaseClient} - returns a 422 if there is already an administrator in the database`, async () => {
        const { app, databaseClient: knex } = await setup({
            databaseClient
        })
    
        const client = Supertest(app)
    
        await createAdminUser(knex)
    
        const response = await client.post('/api/register').send({
            email: 'hey@unknown-user.io',
            password: 'password',
        })
    
        expect(response.status).toBe(422)
        expect(response.body).toMatchSnapshot()
    })
    
    test(`${databaseClient} - correctly creates an administrator user, logs in the user and returns a success message`, async () => {
        const { app } = await setup({
            databaseClient
        })
    
        const client = Supertest(app)
    
        const response = await client.post('/api/register').send({
            email: 'hey@admin.io',
            password: 'password',
            name: 'Hey Admin io',
        })
    
        expect(response.status).toBe(200)
        expect(response.body).toMatchSnapshot()
        expect(response.header['set-cookie']).toHaveLength(1)
    })
    
    test(`${databaseClient} - returns a 200, and creates a new session when correct credentials are passed`, async () => {
        const { app, databaseClient: knex } = await setup({
            databaseClient
        })
 
        const client = Supertest(app)
    
        const user = await createAdminUser(knex)
    
        expect(await knex('sessions').select('*')).toHaveLength(0)
    
        const response = await client.post('/api/login').send({
            email: user.email,
            password: user.password,
        })
    
        expect(response.status).toBe(200)
        expect(response.header['set-cookie']).toHaveLength(1)
    
        expect(response.body).toMatchSnapshot()
    
        const sessions = await knex('sessions').select('*')
    
        expect(sessions).toHaveLength(1)
    
        const session = JSON.parse(sessions[0].sess)
    
        expect(session.user).toBe(user.id)
    })
    
    test(`${databaseClient} - can login correctly with remember me`, async () => {
        const { app, databaseClient: knex } = await setup({
            databaseClient
        })

        const client = Supertest(app)
    
        const user = await createAdminUser(knex)
    
        const response = await client.post('/api/login').send({
            email: user.email,
            password: user.password,
            rememberMe: true,
        })
    
        expect(response.status).toBe(200)
        expect(response.header['set-cookie']).toHaveLength(1)
    
        expect(response.body).toMatchSnapshot()
    
        const sessions = await knex('sessions').select('*')
    
        const session = JSON.parse(sessions[0].sess)
    
        expect(isAfter(new Date(session.cookie.expires), new Date())).toBeTruthy()
    })
    
})