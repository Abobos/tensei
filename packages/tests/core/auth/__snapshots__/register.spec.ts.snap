// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mysql -  validates registration data and returns error messages with a 422 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "The email is required.",
      "validation": "required",
    },
    Object {
      "field": "password",
      "message": "The password is required.",
      "validation": "required",
    },
  ],
  "message": "Validation failed.",
}
`;

exports[`mysql - correctly creates an administrator user, logs in the user and returns a success message 1`] = `
Object {
  "message": "Registration and login successful.",
}
`;

exports[`mysql - returns a 200, and creates a new session when correct credentials are passed 1`] = `
Object {
  "message": "Login successful.",
}
`;

exports[`mysql - returns a 422 if there is already an administrator in the database 1`] = `
Object {
  "message": "An administrator user already exists. Please use the administration management dashboard to add more users.",
}
`;

exports[`pg -  validates registration data and returns error messages with a 422 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "The email is required.",
      "validation": "required",
    },
    Object {
      "field": "password",
      "message": "The password is required.",
      "validation": "required",
    },
  ],
  "message": "Validation failed.",
}
`;

exports[`pg - correctly creates an administrator user, logs in the user and returns a success message 1`] = `
Object {
  "message": "Registration and login successful.",
}
`;

exports[`pg - returns a 200, and creates a new session when correct credentials are passed 1`] = `
Object {
  "message": "Login successful.",
}
`;

exports[`pg - returns a 422 if there is already an administrator in the database 1`] = `
Object {
  "message": "An administrator user already exists. Please use the administration management dashboard to add more users.",
}
`;

exports[`sqlite3 -  validates registration data and returns error messages with a 422 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "The email is required.",
      "validation": "required",
    },
    Object {
      "field": "password",
      "message": "The password is required.",
      "validation": "required",
    },
  ],
  "message": "Validation failed.",
}
`;

exports[`sqlite3 - correctly creates an administrator user, logs in the user and returns a success message 1`] = `
Object {
  "message": "Registration and login successful.",
}
`;

exports[`sqlite3 - returns a 200, and creates a new session when correct credentials are passed 1`] = `
Object {
  "message": "Login successful.",
}
`;

exports[`sqlite3 - returns a 422 if there is already an administrator in the database 1`] = `
Object {
  "message": "An administrator user already exists. Please use the administration management dashboard to add more users.",
}
`;
